src/hashstore/filehashstore.py

```diff
@@ -404,77 +413,237 @@ def lookup_algo(algo):
+            # If no pid is supplied, store the object only without tagging
+            logging.debug("FileHashStore - store_object: Request to store data only.")
+            object_metadata = self.store_data_only(data)
+            logging.info(
+                "FileHashStore - store_object: Successfully stored object for cid: %s",
+                object_metadata.id,
+            )
+        else:
+            # Else the object will be stored and tagged
+            logging.debug(
+                "FileHashStore - store_object: Request to store object for pid: %s", pid
+            )
+            # Validate input parameters
+            self._is_string_none_or_empty(pid, "pid", "store_object")
+            self._validate_arg_data(data)
+            self._is_int_and_non_negative(expected_object_size)
+            (
+                additional_algorithm_checked,
+                checksum_algorithm_checked,
+            ) = self._validate_arg_algorithms_and_checksum(
+                additional_algorithm, checksum, checksum_algorithm
+            )
+
+            # Wait for the pid to release if it's in use
+            while pid in self.object_locked_pids:
+                logging.debug(
+                    "FileHashStore - store_object: %s is currently being stored. Waiting.",
+                    pid,
+                )
+                time.sleep(self.time_out_sec)
+            # Modify object_locked_pids consecutively
+            with self.object_lock:
+                logging.debug(
+                    "FileHashStore - store_object: Adding pid: %s to object_locked_pids.",
+                    pid,
+                )
+                self.object_locked_pids.append(pid)
+            try:
+                logging.debug(
+                    "FileHashStore - store_object: Attempting to store object for pid: %s",
+                    pid,
+                )
+                object_metadata = self.store_and_validate_data(
+                    pid,
+                    data,
+                    additional_algorithm=additional_algorithm_checked,
+                    checksum=checksum,
+                    checksum_algorithm=checksum_algorithm_checked,
+                    file_size_to_validate=expected_object_size,
+                )
+                self.tag_object(pid, object_metadata.id)
+                logging.info(
+                    "FileHashStore - store_object: Successfully stored object for pid: %s",
+                    pid,
+                )
+            finally:
+                # Release pid
+                with self.object_lock:
+                    logging.debug(
+                        "FileHashStore - store_object: Removing pid: %s from object_locked_pids.",
+                        pid,
+                    )
+                    self.object_locked_pids.remove(pid)
+
+        return object_metadata
+
+    def verify_object(
+        self, object_metadata, checksum, checksum_algorithm, expected_file_size
```

Comment:
In the interest of keeping things as simple as possible, do we really need to provide/check `expected_file_size`? If the checksum matches, it's the same file, right?

Reply:
Yes - if the checksum matches, it's the same file. However, after checking in with Jing, it looks like we need to confirm the file size because a quota service exists (which charges users based on the space usage).

> From the technique view, if a checksum match, this means the file is valid. However, we have a quota service, which will charge uses’ money based on the space usage. So we should have check the size of file is correct.